compiler/rustc_mir/src/transform/mod.rs

```diff
@@ -315,6 +317,63 @@ fn mir_promoted(
     (tcx.alloc_steal_mir(body), tcx.alloc_steal_promoted(promoted))
 }
 
+fn mir_for_ctfe<'tcx>(tcx: TyCtxt<'tcx>, def_id: DefId) -> &'tcx Body<'tcx> {
+    let did = def_id.expect_local();
+    if let Some(def) = ty::WithOptConstParam::try_lookup(did, tcx) {
+        tcx.mir_for_ctfe_of_const_arg(def)
+    } else {
+        tcx.arena.alloc(inner_mir_for_ctfe(tcx, ty::WithOptConstParam::unknown(did)))
+    }
+}
+
+fn mir_for_ctfe_of_const_arg<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    (did, param_did): (LocalDefId, DefId),
+) -> &'tcx Body<'tcx> {
+    tcx.arena.alloc(inner_mir_for_ctfe(
+        tcx,
+        ty::WithOptConstParam { did, const_param_did: Some(param_did) },
+    ))
+}
+
+fn inner_mir_for_ctfe(tcx: TyCtxt<'_>, def: ty::WithOptConstParam<LocalDefId>) -> Body<'_> {
+    // FIXME: don't duplicate this between the optimized_mir/mir_for_ctfe queries
+    if tcx.is_constructor(def.did.to_def_id()) {
+        // There's no reason to run all of the MIR passes on constructors when
+        // we can just output the MIR we want directly. This also saves const
+        // qualification and borrow checking the trouble of special casing
+        // constructors.
+        return shim::build_adt_ctor(tcx, def.did.to_def_id());
+    }
+
+    assert_ne!(
+        tcx.hir().body_const_context(def.did),
+        None,
+        "mir_for_ctfe should not be used for runtime functions"
+    );
+
+    let mut body = tcx.mir_drops_elaborated_and_const_checked(def).borrow().clone();
+
+    #[rustfmt::skip]
+    let optimizations: &[&dyn MirPass<'_>] = &[
+        &const_prop::ConstProp,
```

Comment:
Please explain why we still need to run this. Doesn't this lead to duplicate lints (and duplicate work!) for `const fn` as we run the same pass twice on them?

Also doesn't Miri rely on de-aggregation having happened?

Reply:
deaggregation is not an optimization but a required transformation nowadays. Do you think deaggregation an affect UB? I don't think so, assigning element-wise or via a single statement should make no difference. While the miri engine still contains aggregate processing code, I think that is dead code (creating a PR for replacing it with `unreachable!()` right now)

Reply:
Ah, we still have arrays that aren't deaggregated. So no, miri does not rely on deaggregation, but it everything but arrays are deaggreagated.

Reply:
Whatever you want to call it, it's not in this list, so it looks like CTFE will run code that was not deaggregated?

Reply:
CTFE will run deaggregated code, deaggregation happens in an earlier query

Reply:
Oh I did not realize deaggregation was moved. Last time I looked at this file I think it was still in `mir_optimized`.^^